### ISSUE TITLE
[Bug] local_var does not correctly handle BatchedDGLGraph

### ISSUE THREAD

**Issue by yzh119** [dmlc/dgl#739]
## 🐛 Bug

`local_var` would create a new DGLGraph object with the same graph structure of the original graph. However, when the graph is a BatchedDGLGraph, `local_var` would not return a BatchedDGLGraph accordingly. I think for `SubGraph` there is a similar issue.

## To Reproduce

``` python
>>> import dgl
>>> g = dgl.DGLGraph()
>>> g.add_nodes(5)
>>> bg = dgl.batch([g, g, g])
>>> type(bg)
<class 'dgl.batched_graph.BatchedDGLGraph'>
>>> bg = bg.local_var()
>>> type(bg)
<class 'dgl.graph.DGLGraph'>
```

## Expected behavior

`local_var` should return an object of the same class as the caller.

## Environment

 - DGL Version (e.g., 1.0): 0.3
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): 1.1
 - OS (e.g., Linux): mac
 - How you installed DGL (`conda`, `pip`, source): source 
 - Build command you used (if compiling from source): USE_CUDA=OFF
 - Python version: 3.7
 - CUDA/cuDNN version (if applicable): 
 - GPU models and configuration (e.g. V100):
 - Any other relevant information:

## Additional context

`local_scope` does not have such problem because it does not create a new object.

**Comment by jermainewang** [comment]
Yeah, that's a problem. In general, I blame this to having the extra `BatchedGraph` and `Subgraph` classes. In tensor world, there is one data structure `Tensor`. Easy to understand, low learning curve. By contrast, we have so many different classes. This raises issues in many cases. For example, many operations need to check the input type and use different routines (such as in glob.py). My question: could we merge these classes as one and only one `DGLGraph`? It can maintain some extra members such as the `num_nodes` and `num_edges` tensors (if there is only one graph, the tensor is [0]), and also the mapping (none if the graph is not a subgraph). What do you guys think? We could write a RFC to discuss this more.

**Comment by mufeili** [comment]
Yeah, I think to have one class will make the world a better place.

**Comment by VoVAllen** [comment]
I think it's fine to merge `Subgraph` with `DGLGraph`. And `DGLGraph` can be considered as a mutable `BatchedDGLGraph` with batch size 1. So `DGLGraph` will be a subclass of `BatchedDGLGraph`? (not sure, seems a bit weird)
However, my only concern is these designs should be encapsulated within the implementation, and should not break the intuition of the API, i.e. we don't want to educate user `DGLGraph` is a `BatchedDGLGraph` with batch size 1.

**Comment by jermainewang** [comment]
> we don't want to educate user `DGLGraph` is a `BatchedDGLGraph` with batch size 1.

That's a legit concern. Another way is to return these auxiliary data structure to the user so we don't maintain it. In such way, the operations on the BatchedGraph (such as the ones in glob.py) need to be changed to take into these extra arguments.

**Comment by VoVAllen** [comment]
I think can still let DGL hold the auxiliary data structure, and provide a downcast mechanism when needed. Hold auxiliary data sometimes is painful, especially when you want to use them in different scope/function/files.

**Comment by Junyoungpark** [comment]
I guess this issue is still on progress. Does anyone know whether it is actually updated to return `dgl.BatchedDGLgraph` when input is `dgl.BatchedDGLgraph`?

**Comment by yzh119** [comment]
Hi @Junyoungpark I'm gonna fix it soon, I apologize for being lazy.
The `BatchedDGLGraph` class would be deprecated, we would introduce a `batch_size` field for all `DGLGraph`'s and `BatchedDGLGraph` is a special case where `batch_size` is greater than one.

**Comment by yzh119** [comment]
@Junyoungpark I think the problem has been fixed in #1216 . You can try using the latest version of DGL by install from source code in the master branch and see if it works.
